### DISCUSSION TITLE
How to handle shell prompts for colocated repositorys?

### DISCUSSION THREAD

**Discussion by johanneskastl** [jj-vcs/jj#5196] [Q&A]
Hi all,

as the majority of my repos is not yet converted to jj, does anyone have a solution to only show the git prompt in git repos that do **not** yet have a `.jj` repository?

I would like to keep using the git prompt for all those repos instead of disabling it completely...

Kind Regards,
Johannes

**Comment by johanneskastl** [comment]
As a crude hack, I am currently not showing the git prompt if there is a `.jj` directory, using a simple `test`.

```
export PS1="${GREEN}\u${WHITE}@${RED}\H${WHITE}:${BLUE}\${PWD}${VIOLET}\$(test -d .jj > /dev/null || __git_ps1 \" (%s)\")${BLUE}>${WHITE}"
```

I'll see if I can just show the [jj prompt](https://github.com/jj-vcs/jj/wiki/Shell-Prompt) instead.

**Comment by arxanas** [comment]
I'm not sure if this is "blessed", but the best way might be to run `jj root` and check the exit code. (Plus, you'll get the root directory in case you need it for subsequent commands. In your current version, presumably it doesn't work if you're in a subdirectory of the repository?)

**Reply by johanneskastl** [reply]
> presumably it doesn't work if you're in a subdirectory of the repository? 

Of course you are correct. I am currently trying to modify the wiki example. It almost works, but it breaks long lines and word wraps in the terminal. Once I got this solved, I'll post my solution.

**Comment by gasche** [comment]
I ended up looking for a configuration to fix my (bash) prompt when working in a git+jj colocated repository, which currently produces scary output when I am working from `jj`. I found the following online:

- [a jj prompt for power10k (zsh)](https://andre.arko.net/2025/06/20/a-jj-prompt-for-powerlevel10k/)
- [another jj prompt for power10k (zsh)](https://zerowidth.com/2025/async-zsh-jujutsu-prompt-with-p10k/)
- [a jj prompt for zsh](https://github.com/rkh/zsh-jj) (via a synchronous/blocking command rather than asynchronous update)
- a [discussion](https://github.com/fish-shell/fish-shell/issues/11183) of jj prompts for `fish`, which apparently integrates something by default but users in that issue report that they prefer the configuration shared there

Upstream `git` comes with[completion and prompt scripts for common shells: [git/contrib/completion](https://github.com/git/git/tree/master/contrib/completion). It would be convenient for users to have a blessed answer to "how do I fix my prompt to work with git+jj repositories?".

**Reply by gasche** [reply]
Ah, I realize that some of this information is already available on the jj wiki, <https://github.com/jj-vcs/jj/wiki/Shell-Prompt>, apologies for the noise.

**Reply by gasche** [reply]
My previous `~/.bashrc.d/git.sh` file for bash:

```bash
# git completion
source /usr/share/doc/git/contrib/completion/git-completion.bash

GIT_PS1_DESCRIBE_STYLE=branch
source /usr/share/doc/git/contrib/completion/git-prompt.sh
PS1='[\u@\h \W$(__git_ps1 " (%s)")]\$ '
```

My file after taking inspiration from the Wiki:

```bash
# git completion
source /usr/share/doc/git/contrib/completion/git-completion.bash

GIT_PS1_DESCRIBE_STYLE=branch
source /usr/share/doc/git/contrib/completion/git-prompt.sh

__git_prompt() {
    __git_ps1 " (%s)"
}

# https://github.com/jj-vcs/jj/wiki/Shell-Prompt
__jj_prompt() {
    # --ignore-working-copy: avoid inspecting $PWD and concurrent snapshotting which could create divergent commits
    printf '%s' "$(
    jj --ignore-working-copy --no-pager log --no-graph --color=always -r @ -T \
    ' "jj:[@ " ++ concat( separate(" ", format_short_change_id_with_hidden_and_divergent_info(self),format_short_commit_id(commit_id),
    bookmarks,if(conflict,label("conflict","conflict"))))++"]" ' 2>/dev/null
    )"
}


__jjgit_prompt() {
    # echo "jj" or "git" if either is found in $PWD or its parent directories
    pwd_in_jjgit() {
        # using the shell is much faster than `git rev-parse --git-dir` or `jj root`
        local D="/$PWD"
        while test -n "$D"
        do
            test -e "$D/.jj" && { echo jj ; return ; }
            test -e "$D/.git" && { echo git ; return ; }
            D="${D%/*}"
        done
    }

    case "$(pwd_in_jjgit)" in
        jj) (__jj_prompt) ;;
        git) (__git_prompt) ;;
        *) ;;
    esac
}

PS1='[\u@\h \W\[$(__jjgit_prompt)\]]\$ '
```

(Note: I needed to protect the prompt with `\[...\]` around `$(__jjgit_prompt)` to avoid bash miscounting the line width due to color escape codes included in the recommended `jj` prompt.)
